integration/ca/ca.go

```diff
@@ -0,0 +1,123 @@
+package ca
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"os"
+	"time"
+)
+
+type CA struct {
+	key    *ecdsa.PrivateKey
+	cert   *x509.Certificate
+	serial *big.Int
+}
+
+func publicKey(priv interface{}) interface{} {
+	switch k := priv.(type) {
+	case *rsa.PrivateKey:
+		return &k.PublicKey
+	case *ecdsa.PrivateKey:
+		return &k.PublicKey
+	default:
+		return nil
+	}
+}
+
+func New(name string) *CA {
+	key, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+	if err != nil {
+		panic(err)
+	}
+
+	return &CA{
+		key: key,
+		cert: &x509.Certificate{
+			SerialNumber: big.NewInt(1),
+			Subject: pkix.Name{
+				Organization: []string{name},
+			},
+			NotBefore: time.Now(),
+			NotAfter:  time.Now().Add(time.Hour * 24 * 180),
+
+			KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+			BasicConstraintsValid: true,
+			IsCA:                  true,
+		},
+		serial: big.NewInt(2),
+	}
+
+}
+
+func (ca *CA) WriteCACertificate(path string) error {
+	derBytes, err := x509.CreateCertificate(rand.Reader, ca.cert, ca.cert, publicKey(ca.key), ca.key)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return err
+	}
+
+	if err := pem.Encode(f, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes}); err != nil {
+		return err
+	}
+
+	return f.Close()
```

Comment:
I'm not a huge fan of smart one-liners, but this saves lot of boilerplate:

```suggestion
	return ioutil.WriteFile(path, pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: derBytes}), 0644)
```

Similarly in `WriteCertificate` function.

Reply:
I decided to stay with the long form and use O_EXCL too avoid overwriting existing files

Reply:
> I decided to stay with the long form and use O_EXCL too avoid overwriting existing files

Since this code is only used in test code, I don't see overwriting existing files being a problem.

I would still prefer to make code more readable – we can at least extract a `writeExclusiveFile(path, data)` method, as this code is used at several places in the file.

Reply:
OK I have just pushed another update